src/components/vr_cooperation/src/vr_module.cc

```diff
@@ -31,35 +31,243 @@
+  plugin_info_.hmi_function_list.push_back(hmi_api::on_service_deactivated);
+  plugin_info_.hmi_function_list.push_back(hmi_api::on_default_service_chosen);
+}
+
 functional_modules::PluginInfo VRModule::GetPluginInfo() const {
   LOG4CXX_AUTO_TRACE(logger_);
   return plugin_info_;
 }
 
 ProcessResult VRModule::ProcessMessage(application_manager::MessagePtr msg) {
   LOG4CXX_AUTO_TRACE(logger_);
+  DCHECK(msg);
+
+  if (!msg) {
+    LOG4CXX_ERROR(logger_, "Null pointer message received.");
+    return ProcessResult::FAILED;
+  }
+  LOG4CXX_DEBUG(logger_, "Mobile message: " << msg->json_message());
+
+  Json::Value value;
+  Json::Reader reader;
+  reader.parse(msg->json_message(), value);
+
+  std::string function_name;
+
+  // Request or notification
+  if (value.isMember(json_keys::kMethod)) {
+    function_name = value[json_keys::kMethod].asString();
+
+    if (value.isMember(json_keys::kId)) {
+      msg->set_message_type(application_manager::MessageType::kRequest);
+    } else {
+      msg->set_message_type(application_manager::MessageType::kNotification);
+    }
+    // Response
+  } else if (value.isMember(json_keys::kResult)
+             && value[json_keys::kResult].isMember(json_keys::kMethod)) {
+    function_name = value[json_keys::kResult][json_keys::kMethod].asString();
+    msg->set_message_type(application_manager::MessageType::kResponse);
+    // Error response
+  } else if (value.isMember(json_keys::kError)
+             && value[json_keys::kError].isMember(json_keys::kData)
+             && value[json_keys::kError][json_keys::kData].isMember(
+                 json_keys::kMethod)) {
+    function_name = value[json_keys::kError][json_keys::kData]
+                    [json_keys::kMethod].asString();
+    msg->set_message_type(application_manager::MessageType::kErrorResponse);
+  } else {
+    DCHECK(false);
+    return ProcessResult::FAILED;
+  }
+
+  if (value.isMember(json_keys::kId)) {
+    msg->set_correlation_id(value[json_keys::kId].asInt());
+  } else if (application_manager::MessageType::kNotification != msg->type()) {
+    DCHECK(false);
+    return ProcessResult::FAILED;
+  }
+
+  msg->set_protocol_version(application_manager::ProtocolVersion::kV3);
+
+  switch (msg->type()) {
+    case application_manager::MessageType::kResponse:
+    case application_manager::MessageType::kErrorResponse: {
+      if (functional_modules::hmi_api::activate_service == function_name) {
+        VRModuleEvent event(msg, MobileFunctionID::ACTIVATE_SERVICE);
+        EventDispatcher<application_manager::MessagePtr,
+          functional_modules::MobileFunctionID>::instance()->raise_event(event);
+      }
+      return ProcessResult::PROCESSED;
+      break;
+    }
+    case application_manager::MessageType::kNotification: {
+      return ProcessResult::PROCESSED;
+      break;
+    }
+    case application_manager::MessageType::kRequest: {
+      commands::Command* command = MobileCommandFactory::CreateCommand(msg);
+      if (command) {
```

Comment:
may be it is better to check if msg isn't valid abd when return. Using this we will avoid else statement